pkg/cloudprovider/providers/azure/azure_cache.go

```diff
@@ -19,63 +19,138 @@ package azure
 import (
 	"sync"
 	"time"
-
-	"k8s.io/client-go/tools/cache"
 )
 
-type timedcacheEntry struct {
-	key  string
-	data interface{}
-}
+type getFunc func(key string) (interface{}, error)
+type listFunc func() (map[string]interface{}, error)
+
+// timedCache is a cache with TTL.
+type timedCache struct {
+	store       map[string]interface{}
+	lock        sync.Mutex
+	ttl         time.Duration
+	lastRefresh time.Time
 
-type timedcache struct {
-	store cache.Store
-	lock  sync.Mutex
+	getter getFunc
+	lister listFunc
 }
 
-// ttl time.Duration
-func newTimedcache(ttl time.Duration) timedcache {
-	return timedcache{
-		store: cache.NewTTLStore(cacheKeyFunc, ttl),
+// newTimedcache creates a new timedCache with TTL.
+func newTimedcache(ttl time.Duration, getFunc getFunc, listFunc listFunc) *timedCache {
+	return &timedCache{
+		ttl:    ttl,
+		getter: getFunc,
+		lister: listFunc,
+		store:  make(map[string]interface{}),
 	}
 }
 
-func cacheKeyFunc(obj interface{}) (string, error) {
-	return obj.(*timedcacheEntry).key, nil
+// expired returns true if the cache is expired.
+func (t *timedCache) expired() bool {
+	return t.lastRefresh.Add(t.ttl).Before(time.Now())
 }
 
-func (t *timedcache) GetOrCreate(key string, createFunc func() interface{}) (interface{}, error) {
-	entry, exists, err := t.store.GetByKey(key)
-	if err != nil {
+// Get gets data by key. It returns nil if data not found.
+func (t *timedCache) Get(key string) (interface{}, error) {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	if t.expired() {
+		if err := t.refresh(""); err != nil {
+			return nil, err
+		}
+	}
+
+	entry, ok := t.store[key]
+	if ok {
+		return entry, nil
+	}
+
+	// not cache this vm yet, refresh by key
+	if err := t.refresh(key); err != nil {
 		return nil, err
 	}
-	if exists {
-		return (entry.(*timedcacheEntry)).data, nil
+	entry, ok = t.store[key]
+	if ok {
+		return entry, nil
 	}
 
-	t.lock.Lock()
-	defer t.lock.Unlock()
-	entry, exists, err = t.store.GetByKey(key)
+	// Key still not found, set it to nil. This is required to avoid
+	// relisting nonexist objects.
+	t.store[key] = nil
```

Comment:
delete(map, key)

Reply:
Think this as an error processing: if something wrong happened and the object doesn't exist on Azure, with this nil statement, it will trigger GET API on each cache.Get(), even within a TTL expiration. This could help us to avoid unnecessary ARM calls.

Reply:
That shouldn't happen. We need to log the case not hide it.